File: src/DIY/Factory/Factory.java
package DIY.Factory;

public class Factory {
    public static void main(String[] args) {
        WeaponFactory factory = new WeaponFactory();

        Weapon sword = factory.createWeapon("sword");
        Weapon spear = factory.createWeapon("spear");
        Weapon bow = factory.createWeapon("bow");

        sword.attack();
        spear.attack();
        bow.attack();

    }

}

interface Weapon{
    void attack();
}


class Sword implements Weapon{

    @Override
    public void attack() {
        System.out.println("Slashing with a sword...");
    }

}

class Bow implements Weapon{

    @Override
    public void attack() {
        System.out.println("Knocking an arrow...");
    }
    
}

class Spear implements Weapon{

    @Override
    public void attack() {
        System.out.println("Thrusting with a spear...");
    }

}

class WeaponFactory{
    public Weapon createWeapon(String type){
        switch(type.toLowerCase()){
            case "sword":
                return new Sword();
            case "bow":
                return new Bow();
            case "spear":
                return new Spear();
            default:
                throw new IllegalArgumentException("Invalid weapon type.");
        }
    }
}

File: src/DIY/Factory/Factory.ts
// Factory.ts

interface Weapon {
    attack(): void;
}

class Sword implements Weapon {
    attack(): void {
        console.log("Slashing with a sword...");
    }
}

class Bow implements Weapon {
    attack(): void {
        console.log("Knocking an arrow...");
    }
}

class Spear implements Weapon {
    attack(): void {
        console.log("Thrusting with a spear...");
    }
}

class WeaponFactory {
    createWeapon(type: string): Weapon {
        switch (type.toLowerCase()) {
            case "sword":
                return new Sword();
            case "bow":
                return new Bow();
            case "spear":
                return new Spear();
            default:
                throw new Error("Invalid weapon type.");
        }
    }
}

// main
(function main() {
    const factory = new WeaponFactory();

    const sword = factory.createWeapon("sword");
    const spear = factory.createWeapon("spear");
    const bow = factory.createWeapon("bow");

    sword.attack();
    spear.attack();
    bow.attack();
})();
